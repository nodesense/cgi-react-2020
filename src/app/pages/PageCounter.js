// app/pages/PageCounter.js
// useReducer basic

import React, {useReducer} from 'react';

// reducer
// Reducer is a pure function
// reducer doesn't store data
// reducer should SYNC
// reduce SHOULD NOT do async, no API call, no defered call, no promises, no rxjs, no events
// reducer (state, action)
// action is an OBJECT { type: 'STRING preferred/unique one', payload/addtional attributes is completely optional}
// should not mutate the data
// reducers provides logic to action
const reducer = (state, action) => {
    console.log("reducer called, state ", state, " action ", action)
    switch(action.type) {

        case 'increment': 
            //BAD ==> STATE MUTATION, NOT ALLOWED state.counter = state.counter + action.value
            return {counter: state.counter + action.value}

        case 'decrement': 
            return {counter: state.counter - action.value}

        case 'reset': 
            return {counter: 0}

        // TODO Now 'decrement' and 'reset' where no payload to reset

        // have default implementation
        // if no default, not implemented actions would set the state to undefined
        // js by default returns undefined
        default: return state;
    }
}

// dispatch accept action as arg
// dispatch shall call the reducer with current state as first argument, 
// second arg is action

const PageCounter = ({}) => {
    const [state, dispatch] = useReducer(reducer, {counter: 100})
    console.log("State ", state)
    return (
        <div>
            <h2>Page Counter</h2>

            <p>Counter {state.counter} </p>

            <button onClick={() => dispatch({type: 'increment', value: 1}) }> +1 </button>
            <button onClick={() => dispatch({type: 'increment', value: 5}) }> +5 </button>

            <button onClick={() => dispatch({type: 'increment', value: -1}) }> -1 </button>
            <button onClick={() => dispatch({type: 'reset'}) }> Reset </button>
        </div>
    )
}

export default PageCounter;