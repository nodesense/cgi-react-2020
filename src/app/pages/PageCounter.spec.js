// name the test cases either .spec.js or test.js
// JEST shall scan all fiels with spec.js or test.js execute them
import React from 'react';

import PageCounter, {reducer}  from './PageCounter';

import renderer from 'react-test-renderer';
  

// test suite
describe("Testing Reducers Test Suite", () => {
    // a test suite contains more test cases
    // it means test specification
    // english word it.
    it("should increment state value", () => {
        // assert/test statement/predicates
        // expect(1 + 1).toBe(2) // shallow compare [compare two values, or two refs, not deep equal]
        const state = {counter: 0}
        const action = {type: 'increment', value: 10}

        // toEqual , deep equality == without type checking
        // reducer toStrictEqual deep equality === with type checking
        expect(reducer(state, action)).toEqual({counter: 10})
    })

    it("should decrement state value", () => {
        const state = {counter: 100}
        const action = {type: 'decrement', value: 10}
        expect(reducer(state, action)).toEqual({counter: 90})
    })
})

describe("Test PageCouner component test", () => {
    // V.dom testing, NO REAL DOM, no react-dom library, no events/no queue
    // snapshot, convert V DOM into JSON format
    it("test page counter defaults", () => {
        // create component, render it to V.DOM
        // convert to JSON tree
        const tree = renderer
                    .create(<PageCounter />)
                    .toJSON();

        // very first time, it will create a folder, create a reference copy for test
        // for subsequest, test, reference copy used for comparison
        expect(tree).toMatchSnapshot();
    })
})