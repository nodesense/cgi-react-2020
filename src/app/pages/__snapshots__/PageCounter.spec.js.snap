// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PageCouner component test test page counter defaults 1`] = `
<div>
  <h2>
    Page Counter
  </h2>
  <p
    className="p1"
  >
    Counter 
    100
     
  </p>
  <button
    onClick={[Function]}
  >
     +1 
  </button>
  <button
    onClick={[Function]}
  >
     +5 
  </button>
  <button
    onClick={[Function]}
  >
     -1 
  </button>
  <button
    onClick={[Function]}
  >
     Reset 
  </button>
</div>
`;
